File: Ford_Fulkerson_Method/Ford_Fulkerson_Method.js
//+++++++++++++++Utility+++++++++++++++++
//Queue Data Structure implemented 
function Queue() {
    this.elements = [];
 }
 Queue.prototype.enqueue = function (e) {
    this.elements.push(e);
 };
 // remove an element from the front of the queue
Queue.prototype.dequeue = function () {
    return this.elements.shift();
};
// check if the queue is empty
Queue.prototype.isEmpty = function () {
    return this.elements.length == 0;
};
// get the element at the front of the queue
Queue.prototype.peek = function () {
    return !this.isEmpty() ? this.elements[0] : undefined;
};
Queue.prototype.length = function() {
    return this.elements.length;
}

//++++++++++++++++++++++++++++++++++++++++++
//Function to do breadth first search in graph 
function bfs(reducedGraph, source, sink, path, V)
{
    //array to store if the node is visited or not 
    var visited = [V];
    //intially no node is visited 
    for (var i = 0; i < V; i++)
        visited[i] = false;
    let q = new Queue();
    //Starting from source
    q.enqueue(source);
    visited[source] = true;
    //parent of starting node will be nothing hence -1 is stored in it
    path[source] = -1;

    while (!q.isEmpty())
    {
        var u = q.peek();
        q.dequeue();
        for(var v = 0; v < V; v++)
        {
            if(visited[v] === false && reducedGraph[u][v] > 0)
            {
                q.enqueue(v);
                path[v] = u;
                visited[v] = true;
            }
        }
    }
    //If there exist path from source to sink then it will return true else false
    return (visited[sink] === true);
}

function fordFulkerson(graph,source,sink,V)
{
    //reducedGraph is used to store the remaining capacity. reducedGraph[i][j] 
    //indicates remaining capacity of edge from i to jif edge exists.
    var reducedGraph = [];
    for( var i = 0; i < V; i++)
    {
        reducedGraph[i] = [];
        for(var j = 0; j < V; j++)
        {
            reducedGraph[i][j] = graph[i][j];
        }
    }
    //Path is used to store parent  
    var path = [V];
    //variable to store maximum flow in the path chosen
    var max_flow = 0;
    while (bfs(reducedGraph, source, sink, path, V))
    {
        var path_flow = Number.MAX_VALUE;
        //Finding maximum flow in the path selected
        for(var v = sink; v != source; v = path[v])
        {
            var u = path[v];
            path_flow = Math.min( path_flow, reducedGraph[u][v]);
        }
        //Updating remaining capacity odf the edges
        for( var v = sink ; v != source; v = path[v])
        {
            var u = path[v];
            reducedGraph[u][v] -= path_flow;
            reducedGraph[v][u] += path_flow;
        }
        max_flow += path_flow;
    }
    return max_flow;
}
//Used synchronous readline package to take console input 
var readlineSync = require('readline-sync');
//Input number of nodes in a graph
var V = readlineSync.question('Enter number of nodes in a graph: ');
//Input the weight matrix for the directed graph

var graph = [];
for(var i = 0; i < V; i++)
{
    graph[i] = [];
    for(var j = 0; j < V; j++)
    {
        console.log(`Enter the weight form ` , i , `to `,j, `: `);
        var a = readlineSync.question();
        graph[i][j] = a;
    }
}
var source = readlineSync.question("Enter source: ");
var sink = readlineSync.question("Enter Sink: ");
var ans = fordFulkerson(graph,source,sink,V);
console.log('Maximum flow from source to sink is: ', ans);
/*Input1:
V = 6
graph = [ [0, 16, 13, 0, 0, 0], 
        [0, 0, 10, 12, 0, 0], 
        [0, 4, 0, 0, 14, 0], 
        [0, 0, 9, 0, 0, 20], 
        [0, 0, 0, 7, 0, 4], 
        [0, 0, 0, 0, 0, 0] 
      ];
source = 0 sink = 5
Output1: 23

Input2:
V = 6
graph = [ [0, 10, 0, 3, 0, 0], 
        [0, 0, 8, 2, 0, 0], 
        [0, 0, 0, 0, , 10], 
        [0, 0, 6, 0, 7, 0], 
        [0, 0, 0, 0, 10, 0], 
        [0, 0, 0, 0, 0, 0] 
      ];
source = 0
sink = 5
Output2: 10
*/

File: Ford_Fulkerson_Method/Ford_Fulkerson_Method.c
#include <stdio.h>
#define WHITE 0
#define GRAY 1
#define BLACK 2
#define MAX_NODES 1000
#define oo 1000000000

int n;  // number of nodes
int e;  // number of edges
int capacity[MAX_NODES][MAX_NODES]; // capacity matrix
int flow[MAX_NODES][MAX_NODES];     // flow matrix
int color[MAX_NODES]; // needed for breadth-first search               
int pred[MAX_NODES];  // array to store augmenting path

int min (int x, int y) {
    return x < y ? x : y;  // returns minimum of x and y
}

int head, tail;
int q[MAX_NODES + 2];

void enqueue (int x) {
    q[tail] = x;
    tail++;
    color[x] = GRAY;
}

int dequeue () {
    int x = q[head];
    head++;
    color[x] = BLACK;
    return x;
}

int bfs (int start, int target) {
    int u,v;
    for (u = 0; u < n; u++) {
	color[u] = WHITE;
    }   
    head = tail = 0;
    enqueue(start);
    pred[start] = -1;
    while (head != tail) {
	u = dequeue();
        // Search all adjacent white nodes v. If the capacity
        // from u to v in the residual network is positive,
        // enqueue v.
	for (v = 0; v < n; v++) {
	    if (color[v] == WHITE && capacity[u][v] - flow[u][v] > 0) {
		enqueue(v);
		pred[v] = u;
	    }
	}
    }
    // If the color of the target node is black now,
    // it means that we reached it.
    return color[target] == BLACK;
}

int max_flow (int source, int sink) {
    int i, j, u;
    // Initialize empty flow.
    int max_flow = 0;
    for (i = 0; i < n; i++) {
	for (j = 0; j < n; j++) {
	    flow[i][j] = 0;
	}
    }
    // While there exists an augmenting path,
    // increment the flow along this path.
    while (bfs(source, sink)) {
        // Determine the amount by which we can increment the flow.
	int increment = oo;
	for (u = n - 1; pred[u] >= 0; u = pred[u]) {
	    increment = min(increment, capacity[pred[u]][u] - flow[pred[u]][u]);
	}
        // Now increment the flow.
	for (u = n - 1; pred[u] >= 0; u = pred[u]) {
	    flow[pred[u]][u] += increment;
	    flow[u][pred[u]] -= increment;
	}
	max_flow += increment;
    }
    // No augmenting path anymore. We are done.
    return max_flow;
}

//Reading the input file and the main program

void read_input_file() {
    int a, b, c, i, j;
    FILE* input = fopen("mf.in", "r");
    // read number of nodes and edges
    fscanf(input, "%d %d", &n, &e);
    // initialize empty capacity matrix 
    for (i = 0; i < n; i++) {
	for (j = 0; j < n; j++) {
	    capacity[i][j] = 0;
	}
    }
    // read edge capacities
    for (i=0; i<e; i++) {
	fscanf(input, "%d %d %d", &a, &b, &c);
	capacity[a][b] = c;
    }
    fclose(input);
}

int main () {
    read_input_file();
    printf("%d\n", max_flow(0, n - 1));
    return 0;
}

